Source/WebKit/Shared/ios/HardwareKeyboardState.cpp

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "HardwareKeyboardState.h"
+
+#if PLATFORM(IOS_FAMILY)
+
+#import <pal/spi/ios/GraphicsServicesSPI.h>
+
+namespace WebKit {
+
+HardwareKeyboardState currentHardwareKeyboardState()
+{
+    HardwareKeyboardState state;
+    state.isAttached = GSEventIsHardwareKeyboardAttached();
+    state.countryCode = GSEventGetHardwareKeyboardCountry();
+    state.keyboardType = GSEventGetHardwareKeyboardType();
+    return state;
```

Comment:
Optional suggestion for brevity: If you want to keep the fields explicit, maybe this?
```suggestion
    return {
        .isAttached = GSEventIsHardwareKeyboardAttached(),
        .countryCode = GSEventGetHardwareKeyboardCountry(),
        .keyboardType = GSEventGetHardwareKeyboardType(),
    };
```

Or just drop the designated initializers and go with `return { ..., ..., ... }`

Reply:
Making either change ends up introducing this error:

```
error: non-constant-expression cannot be narrowed from type 'Boolean' (aka 'unsigned char') to 'bool' in initializer list [-Wc++11-narrowing]
```

Keeping as-is, unless the explicit cast is also preferred?